### PR TITLE
[dynamo] Support object creation of classes with custom __new__

### PR DESCRIPTION
**Pull request by anijain2305**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #132977

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @kadeng @chauhang @amjames

### CODE REVIEW

File: torch/_dynamo/variables/user_defined.py [hunk 1]
```diff
@@ -489,6 +489,18 @@ def call_function(
         elif issubclass(self.value, enum.Enum) and len(args) == 1 and not kwargs:
             options = {"mutable_local": MutableLocal()}
             return variables.EnumVariable.create(self.value, args[0], options)
+        elif (
+            not self.is_standard_new()
+            and SideEffects.cls_supports_mutation_side_effects(self.value)
```
Comment (anijain2305) [inline torch/_dynamo/variables/user_defined.py:494]: I want to relax this constraint in a future PR. I don't understand why we need this yet.

**Review by jansel — APPROVED** [review]

**Comment by anijain2305** [comment]
Cancelling the merge command to test internally first given @ezyang observations.

File: torch/_dynamo/variables/user_defined.py (hunk 1 above)
Comment (ezyang) [reply torch/_dynamo/variables/user_defined.py:500]: So why does this constraint exist?

Comment (anijain2305) [reply torch/_dynamo/variables/user_defined.py:500]: I could not get a chance to follow up. I remember seeing many tests failing. I can follow up on this now.

**Pull request closed without merging**